### PR TITLE
Reduce FuncDeclaration size

### PR DESCRIPTION
**Pull request by RazvanN7**
by grouping contract-related member fields.

### CODE REVIEW

File: compiler/src/dmd/func.d [hunk 1]
```diff
@@ -258,22 +258,29 @@ extern (C++) struct Ensure
 
 }
 
-/***********************************************************
- */
-extern (C++) class FuncDeclaration : Declaration
+// Most function don't have contracts, so save memory by grouping
```
Comment (maxhaton) [inline compiler/src/dmd/func.d:261]:
```suggestion
// Most functions don't have contracts, so save memory by grouping
```

Also use a multiline doc comment

File: compiler/src/dmd/func.d [hunk 2]
```diff
@@ -258,22 +258,29 @@ extern (C++) struct Ensure
...
     FuncDeclaration fdrequire;          /// function that does the in contract
     FuncDeclaration fdensure;           /// function that does the out contract
-
     Expressions* fdrequireParams;       /// argument list for __require
     Expressions* fdensureParams;        /// argument list for __ensure
+}
+
+/***********************************************************
+ */
+extern (C++) class FuncDeclaration : Declaration
```
Comment (maxhaton) [inline compiler/src/dmd/func.d:277]: This diff looks unnecessary. The class declaration should be unaffected

File: compiler/src/tests/cxxfrontend.cc [hunk 3]
```diff
@@ -964,8 +964,8 @@ class MiniGlueVisitor : public Visitor
         else
             sym = func->searchLabel(s->ident, s->loc);
         sym->statement->accept(this);
-        if (sym == func->returnLabel && func->fensure != NULL)
-            func->fensure->accept(this);
+        if (sym == func->returnLabel && func->fensure() != NULL)
```
Comment (maxhaton) [inline compiler/src/tests/cxxfrontend.cc:967]: This is probably a breaking change for ldc and gdc glue code?

@ibuclaw @kinke ?

**Review by maxhaton — APPROVED** [review]

**Review by maxhaton — CHANGES_REQUESTED** [review]
Previous approve was by mistake, sorry.

**Comment by maxhaton** [comment]
frontend.h!

**Comment by maxhaton** [comment]
frontend.h!

File: compiler/src/dmd/func.d [hunk 4]
```diff
@@ -403,6 +410,75 @@ extern (C++) class FuncDeclaration : Declaration
...
+        inout(Expressions*) fdrequireParams() inout { return contracts ? contracts.fdrequireParams: null; }
+        inout(Expressions*) fdensureParams() inout { return contracts ? contracts.fdensureParams: null; }
+
+        // setters
+        Statements* frequires(Statements* frs)
+        {
+            if (!frs && !contracts)
+                return null;
+            return getContracts().frequires = frs;
+        }
```
Comment (WalterBright) [inline compiler/src/dmd/func.d:438]: With a lot of duplicated code, a mixin template might be more suitable. Or a string mixin.

**Review by WalterBright — APPROVED** [review]

File: compiler/src/dmd/func.d (hunk 2 above)
Comment (RazvanN7) [reply compiler/src/dmd/func.d:277]: Github is just being weird here, I didn't modify the class definition.

**Comment by RazvanN7** [comment]
@maxhaton I addressed your concerns.

@WalterBright Yes, that is what I thought initially, but I wanted to get it right first (as string mixins are a bit cumbersome to debug) and see if it is something that will get accepted. I am using a string mixin now.

**Comment by RazvanN7** [comment]
@maxhaton is this good to go?

**Comment by maxhaton** [comment]
I'm happy with it, I assume Iain and Martin can work around it but wait for confirmation from them (ldc is currently lagging behind dmd, this isn't dramatic breakage if it is indeed breakage but just keep it in mind)

**Review by maxhaton — APPROVED** [review]

**Comment by kinke** [comment]
Yeah this is just regular breakage which the C++ compilers handle quite well usually; additionally, I doubt these members are used in the glue layer.

**Review by thewilsonator — APPROVED** [review]

**Comment by RazvanN7** [comment]
@kinke Is there anything I can do to make things easier for you?

**Pull request merged**